### DISCUSSION TITLE
Bankable CCAbsoluteEncoder

### DISCUSSION THREAD

**Discussion by raychmond** [tttapa/Control-Surface#1029] [Q&A]
Hi, I'm working on the Adafruit MacroPad RP2040. As you know, it has one encoder and 12 keys. What I wanted is to use keys to select the CC address and the encoder to make control changes. Here is my code:
```
Bank<4> bank;
IncrementDecrementSelector<4> selector {
  bank,       // Bank to manage
  {10, 12},     // push button pins (increment, decrement)
  Wrap::Wrap, // Wrap around
};

Bankable::CCAbsoluteEncoder enc {
  {bank, BankType::ChangeAddress},      // bank configuration
  {17, 18},                                   //encoder pins
  {0x01}, // base address
  1,
  4,
};

```
I don't find any examples about the bankable CCAbsoluteEncoder. I'm trying to modify the example of the bankable CCPotentiometer.  But it doesn't allow me to compile. 
I'm still learning this library, especially the bankable. 
Thanks for your help!

**Comment by tttapa** [comment]
Please post your full sketch, and include the full error message.

The number of banks for the encoder is missing: `Bankable::CCAbsoluteEncoder<4>`

**Comment by raychmond** [comment]
Amazing! It worked! Plus, it remembers the last position of each CC even after I switched to different addresses. 
I just started the coding. Nothing there yet...
One more question please, what if I wanted to select from a list of addresses? I know you have mentioned in the documentation. But I don't understand how to implement.

Thank you!

```
#include <Control_Surface.h>

USBMIDI_Interface midi;

Bank<4> bank;
IncrementDecrementSelector<4> selector {
  bank,       // Bank to manage
  {10, 12},     // push button pins (increment, decrement)
  Wrap::Wrap, // Wrap around
};

Bankable::CCAbsoluteEncoder<4> enc {
  {bank, BankType::ChangeAddress},      // bank configuration
  {17, 18},                                   //encoder pins
  {0x01}, // base address
  1,
  4,
};

void setup() {
    Control_Surface.begin(); // Initialize Control Surface
}

void loop() {    
    Control_Surface.loop(); // Update the Control Surface
} 
```

**Reply by tttapa** [reply]
You'll have to use `ManyAddresses::CCAbsoluteEncoder` in that case: https://tttapa.github.io/Control-Surface-doc/Doxygen/da/d5c/classBankable_1_1ManyAddresses_1_1CCAbsoluteEncoder.html#af0fd0fa95a27eb0f124cc1158fe3f8e1

**Reply by raychmond** [reply]
Thank you so much for the quick answer. But I still can't figure it out. Here is my code:

```
#include <Control_Surface.h>

USBMIDI_Interface midi;

Bank<4> bank;
IncrementDecrementSelector<4> selector {
  bank,       // Bank to manage
  {10, 12},     // push button pins (increment, decrement)
  Wrap::Wrap, // Wrap around
};

Bankable::ManyAddresses::CCAbsoluteEncoder<4> enc[] {
  bank,       
  {17, 18},
  {0x01,0x03,0x07,0x09},
  1,
  4,
};

void setup() {
    Control_Surface.begin(); // Initialize Control Surface
}

void loop() {    
    Control_Surface.loop(); // Update the Control Surface
} 
```
I'm hoping to control these four addresses (1,3,7,9) instead of (1,2,3,4) using one encoder.

Here is the error message:
`Compilation error: could not convert 'bank' from 'cs::Bank<4>' to 'cs::Bankable::ManyAddresses::CCAbsoluteEncoder<4>'`

**Reply by raychmond** [reply]
@tttapa Never mind. I removed the brackets after the enc[]. It works now. I don't know why I added that brackets.
